README.md

```
  __  __      _    _____  _____      _      
 |  \/  |    | |  / ____|/ ____|    | |     
 | \  / | ___| |_| (___ | |     __ _| | ___ 
 | |\/| |/ _ \ __|\___ \| |    / _` | |/ __|
 | |  | |  __/ |_ ____) | |____ (_| | | (__ 
 |_|  |_|\___|\__|_____/ \_____\__,_|_|\___|
```                                          
 
Metabolic Syndrome Severity Calculator
======================================

[![DOI](https://zenodo.org/badge/159721335.svg)](https://zenodo.org/badge/latestdoi/159721335)

MetS Calc, the metabolic syndrome (MetS) calculator, is an HTML and JavaScript
implementation that calculates someone's Metabolic Syndrome Severity Score.

 * Website: https://metscalc.org/
 * Source code: https://github.com/metscalc/metscalc

Cite Us
----------

Please reference the investigators’ article(s) below as well as the Clinical
and Translational Science - Informatics and Technology group (CTS-IT) in any
research report, journal, or publication that requires citation. Recognition of
CTS-IT resources you used to perform research is important for acquiring
funding for the next generation of informatics services and our research and
development activities in software development and information science.

If you used the child score, please cite the following article:
https://www.ncbi.nlm.nih.gov/pubmed/23062212

If you used the adult score, please cite the following article:
https://www.ncbi.nlm.nih.gov/pubmed/24290837

### Cite child score

Gurka, M. J., Ice, C. L., Sun, S. S., & DeBoer, M. D. (2012). A confirmatory factor analysis of the metabolic syndrome in adolescents: an examination of sex and racial/ethnic differences. Cardiovascular diabetology, 11(1), 128.

### Cite adult score

Gurka, M. J., Lilly, C. L., Oliver, M. N., & DeBoer, M. D. (2014). An examination of sex and racial/ethnic differences in the metabolic syndrome among adults: a confirmatory factor analysis and a resulting continuous severity score. Metabolism, 63(2), 218-225.

### Cite software

Gurka, M. J., DeBoer, M. D., Filipp, S. L., Khan, J. Z., Rapczak, T. J., Braun, N. D., Hanson, K. S., & Barnes, C. P. (2019). MetS Calc: Metabolic Syndrome Severity Calculator. doi: 10.5281/zenodo.2542213


Quickstart
----------

    $ make
    $ open www/index.html


Running Tests
-------------

    $ make test


ASCII Logo Thanks to : http://textart.io/figlet ( font style "big" )


CHANGELOG
Changelog
=========

[1.0.0] 2019-01-31

 * Removed max bounds.
 * Added unit tests and Travis CI integration.
 * Removed erroneous percent sign.
 * Updated transpilation for IE11 support.

[0.1.0-alpha3] 2019-01-23

 * Improve documentation

[0.1.0-alpha2] 2019-01-23

 * Add max bound and check validity of fields
 * Enable Calculate always and report errors on click
 * Update documentation with canonical URL
 * Improve look on mobile devices
 * Add percentiles for MetS and BMI z-scores
 * Limit display of BMI and BMI Z-Score to 1000ths
 * Use US Government standard for lb-kg conversion

[0.1.0-alpha]

 * Initial version



tests/bmi_test.js
/*
   Copyright 2019 University of Florida

   Licensed under the Apache License, Version 2.0 (the "License");
   you may not use this file except in compliance with the License.
   You may obtain a copy of the License at

       http://www.apache.org/licenses/LICENSE-2.0

   Unless required by applicable law or agreed to in writing, software
   distributed under the License is distributed on an "AS IS" BASIS,
   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
   See the License for the specific language governing permissions and
   limitations under the License.
*/

const bmi = require('../src/bmi')
const test = require('tape')

function nearlyEqual (a, b, margin = 0.001) {
  return a - b < margin
}

test('test boy 120 months', function (t) {
  const age = 10 * 12
  const sex = bmi.Sex.Male
  const weight = 30 // kg
  const height = 1.36 // m
  t.true(nearlyEqual(bmi.BMIZscore(weight, height, sex, age), -0.22, 0.001))
  t.end()
})

test('test boy 120.1 months', function (t) {
  const age = 10.1 * 12
  const sex = bmi.Sex.Male
  const weight = 30 // kg
  const height = 1.36 // m
  t.true(nearlyEqual(bmi.BMIZscore(weight, height, sex, age), -0.22, 0.001))
  t.end()
})



tests/msscalc_test.js
/*
   Copyright 2019 University of Florida

   Licensed under the Apache License, Version 2.0 (the "License");
   you may not use this file except in compliance with the License.
   You may obtain a copy of the License at

       http://www.apache.org/licenses/LICENSE-2.0

   Unless required by applicable law or agreed to in writing, software
   distributed under the License is distributed on an "AS IS" BASIS,
   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
   See the License for the specific language governing permissions and
   limitations under the License.
*/

const bmi = require('../src/bmi')
const msscalc = require('../src/msscalc')
const test = require('tape')

const cases = [
  {
    bmiZScore: 1.3388313623928048,
    'hdl': 47,
    'height': 143.5,
    'sex': 'MALE',
    'age': 13,
    'sbp': 110,
    'race': 'HISPANIC',
    'mets_z_bmi': 0.488099,
    mets_z_wc: '',
    'waist': 76.4,
    'triglyceride': 94,
    'weight': 48,
    'glucose': 93.07662
  },
  {
    bmiZScore: 0.32010601882928535,
    'glucose': 102.9165,
    'triglyceride': 52,
    'weight': 58.1,
    'sbp': 107.3333,
    'race': 'HISPANIC',
    'mets_z_bmi': '-0.33964',
    mets_z_wc: '',
    'waist': 87.5,
    'height': 160.1,
    'hdl': 54,
    'sex': 'FEMALE',
    'age': 19
  },
  {
    bmiZScore: 0.6803390791459352,
    'glucose': 100.6072,
    'triglyceride': 76,
    'weight': 65,
    'waist': 76.5,
    mets_z_wc: '',
    'mets_z_bmi': 0.38623,
    'race': 'WHITE',
    'sbp': 108,
    'age': 15,
    'sex': 'MALE',
    'hdl': 40,
    'height': 171.9
  },
  {
    bmiZScore: 0.9794850777865289,
    'mets_z_bmi': '-0.39342',
    mets_z_wc: '',
    'waist': 90,
    'sbp': 100.6667,
    'race': 'BLACK',
    'sex': 'FEMALE',
    'age': 18,
    'hdl': 68,
    'height': 167.7,
    'glucose': 67.67353,
    'triglyceride': 103,
    'weight': 71.3
  },
  {
    'age': 23,
    'sex': 'FEMALE',
    'hdl': 43,
    'height': 158.9,
    'waist': 81,
    'mets_z_wc': '-0.07672',
    'mets_z_bmi': '-0.0238',
    'race': 'HISPANIC',
    'sbp': 102.6667,
    'weight': 59.8,
    'triglyceride': 117,
    'glucose': 95.28558
  },
  {
    'race': 'WHITE',
    'sbp': 119.3333,
    'waist': 110.4,
    'mets_z_wc': 0.274626,
    'mets_z_bmi': 0.164624,
    'hdl': 46,
    'height': 172.6,
    'age': 39,
    'sex': 'MALE',
    'glucose': 110.6479,
    'weight': 88.6,
    'triglyceride': 97
  },
  {
    'glucose': 107.8365,
    'weight': 83.3,
    'triglyceride': 108,
    'mets_z_wc': 0.180919,
    'mets_z_bmi': 0.228039,
    'waist': 83.6,
    'sbp': 132.6667,
    'race': 'BLACK',
    'sex': 'MALE',
    'age': 22,
    'height': 193.5,
    'hdl': 36
  },
  {
    'age': 43,
    'sex': 'MALE',
    'height': 170.9,
    'hdl': 45,
    'waist': 88.1,
    'mets_z_wc': '-0.50851',
    'mets_z_bmi': '-0.47154',
    'race': 'HISPANIC',
    'sbp': 113.3333,
    'triglyceride': 62,
    'weight': 70.5,
    'glucose': 100.306
  },
  {
    'glucose': 91.77132,
    'triglyceride': 249,
    'weight': 71.2,
    'mets_z_bmi': 0.142653,
    'mets_z_wc': 0.43294,
    'waist': 102.4,
    'sbp': 117.3333,
    'race': 'WHITE',
    'sex': 'FEMALE',
    'age': 58,
    'height': 162.7,
    'hdl': 71
  },
  {
    'sbp': 118,
    'race': 'BLACK',
    'mets_z_bmi': '-0.45297',
    'mets_z_wc': '-0.44722',
    'waist': 88.6,
    'height': 178.9,
    'hdl': 51,
    'sex': 'MALE',
    'age': 44,
    'glucose': 100.5068,
    'triglyceride': 65,
    'weight': 79.8
  },
  {
    'weight': 72.1,
    'triglyceride': 210,
    'glucose': 146.7946,
    'age': 61,
    'sex': 'FEMALE',
    'height': 176.3,
    'hdl': 66,
    'waist': 91,
    'mets_z_bmi': 1.289268,
    'mets_z_wc': 1.371423,
    'race': 'HISPANIC',
    'sbp': 110.6667
  },
  {
    'triglyceride': 181,
    'weight': 95,
    'glucose': 107.6357,
    'height': 176,
    'hdl': 38,
    'age': 85,
    'sex': 'MALE',
    'race': 'WHITE',
    'sbp': 168,
    'waist': 119.7,
    'mets_z_wc': 1.302729,
    'mets_z_bmi': 1.09538
  },
  {
    'glucose': 114.0618,
    'triglyceride': 125,
    'weight': 68.2,
    'waist': 99.4,
    'mets_z_wc': 0.906364,
    'mets_z_bmi': 0.822064,
    'race': 'BLACK',
    'sbp': 135,
    'age': 78,
    'sex': 'FEMALE',
    'hdl': 39,
    'height': 159.2
  }
].map(data => Object.assign(data, {
  bmi: bmi.BMIAdult(data.weight, data.height / 100)
}))

function nearlyEqual (expected, actual, errorMargin = 0.0025) {
  return Math.abs(expected - actual) < errorMargin
}

cases.forEach(data => {
  const testname = `test ${data.age < 20 ? 'CHILD' : 'ADULT'} ${data.race} ${data.sex}`
  test(testname, function (t) {
    const results = msscalc.CalculateMSS(data)
    if (data.mets_z_wc === '') {
      t.true(!results.mets_z_wc, 'No MetS WC for child')
    } else {
      t.true(nearlyEqual(data.mets_z_wc, results.mets_z_wc), `${data.mets_z_wc} ~= ${results.mets_z_wc}`)
    }

    t.true(nearlyEqual(data.mets_z_bmi, results.mets_z_bmi), `${data.mets_z_bmi} ~= ${results.mets_z_bmi}`)
    t.end()
  })
})


src/cli.js
/*
   Copyright 2019 University of Florida

   Licensed under the Apache License, Version 2.0 (the "License");
   you may not use this file except in compliance with the License.
   You may obtain a copy of the License at

       http://www.apache.org/licenses/LICENSE-2.0

   Unless required by applicable law or agreed to in writing, software
   distributed under the License is distributed on an "AS IS" BASIS,
   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
   See the License for the specific language governing permissions and
   limitations under the License.
*/

/* METS Calculator
 * Command Line Interface
 *
 * TODO: better argument handling
 */

const path = require('path')
const msscalc = require('./msscalc')

function main () {
  let [, progname, ...args] = process.argv
  progname = path.basename(progname)

  // age, sex, race, bmi, hdl, sbp, triglyceride, glucose, bmiZScore, waist
  const details = {}

  if (args.length < 1) {
    console.log('Usage: node', progname, 'age=<age> sex=FEMALE|MALE race=BLACK|HISPANIC|WHITE bmi=<bmi> hdl=<hdl> sbp=<sbp> triglyceride=<tri> glucose=<glucose> bmiZScore=<bmiz> waist=<waist>')
    console.log('\nCalculates the Metabolic Syndrome Severity score based on the given inputs.\n')
    console.log('All the arguments are case-sensitive and most of their values')
    console.log('must be numbers (except sex and race).')
    return
  }

  args.forEach(pair => {
    const [key, value] = pair.split('=')
    const num = parseFloat(value)
    details[key] = isNaN(num) ? value : num
  })

  // console.log("%j", details)
  console.log(msscalc.CalculateMSS(details))
}

main()


src/msscalc.js
/*
   Copyright 2019 University of Florida

   Licensed under the Apache License, Version 2.0 (the "License");
   you may not use this file except in compliance with the License.
   You may obtain a copy of the License at

       http://www.apache.org/licenses/LICENSE-2.0

   Unless required by applicable law or agreed to in writing, software
   distributed under the License is distributed on an "AS IS" BASIS,
   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
   See the License for the specific language governing permissions and
   limitations under the License.
*/

const assert = require('assert')

const RaceEthnicity = {
  Black: 'BLACK',
  Hispanic: 'HISPANIC',
  White: 'WHITE'
}

const Sex = {
  Female: 'FEMALE',
  Male: 'MALE'
}

function CalculateMSS (args) {
  const { log } = Math

  assert(args, 'missing all arguments')

  let { age, sex, race, bmi, hdl, sbp, triglyceride, glucose, bmiZScore, waist } = args

  age = parseInt(args.age)
  assert(age >= 0, 'age must be a non-negative integer')
  assert(Object.keys(Sex).some(s => Sex[s] === sex), 'sex must be FEMALE or MALE')
  assert(Object.keys(RaceEthnicity).some(r => RaceEthnicity[r] === race), 'race must be BLACK, HISPANIC, or WHITE')

  if (age >= 20) {
    /* Adults */
    if (sex === Sex.Female) {
      switch (race) {
        case RaceEthnicity.Black:
          return {
            mets_z_bmi: bmi ? -6.7982 + 0.0484 * bmi - 0.0108 * hdl + 0.0073 * sbp + 0.5278 * log(triglyceride) + 0.0281 * glucose : null,
            mets_z_wc: waist ? (-7.1913 + 0.0304 * waist - 0.0095 * hdl + 0.0054 * sbp + 0.4455 * log(triglyceride) + 0.0225 * glucose) : null
          }
        case RaceEthnicity.Hispanic:
          return {
            mets_z_bmi: bmi ? -7.1844 + 0.0333 * bmi - 0.0166 * hdl + 0.0085 * sbp + 0.8625 * log(triglyceride) + 0.0221 * glucose : null,
            mets_z_wc: waist ? (-7.7641 + 0.0162 * waist - 0.0157 * hdl + 0.0084 * sbp + 0.8872 * log(triglyceride) + 0.0206 * glucose) : null
          }
        case RaceEthnicity.White:
          return {
            mets_z_bmi: bmi ? -6.5231 + 0.0523 * bmi - 0.0138 * hdl + 0.0081 * sbp + 0.6125 * log(triglyceride) + 0.0208 * glucose : null,
            mets_z_wc: waist ? (-7.2591 + 0.0254 * waist - 0.0120 * hdl + 0.0075 * sbp + 0.5800 * log(triglyceride) + 0.0203 * glucose) : null
          }
      }
    }

    /* Male */
    switch (race) {
      case RaceEthnicity.Black:
        // Adult Male Black Non-Hispanic
        return {
          mets_z_bmi: bmi ? -4.8134 + 0.0460 * bmi - 0.0233 * hdl + 0.0020 * sbp + 0.5983 * log(triglyceride) + 0.0166 * glucose : null,
          mets_z_wc: waist ? (-6.3767 + 0.0232 * waist - 0.0175 * hdl + 0.0040 * sbp + 0.5400 * log(triglyceride) + 0.0203 * glucose) : null
        }
      case RaceEthnicity.Hispanic:
        // Adult Male Hispanic
        return {
          mets_z_bmi: bmi ? -4.8198 + 0.0355 * bmi - 0.0303 * hdl + 0.0051 * sbp + 0.7835 * log(triglyceride) + 0.0104 * glucose : null,
          mets_z_wc: waist ? (-5.5541 + 0.0135 * waist - 0.0278 * hdl + 0.0054 * sbp + 0.8340 * log(triglyceride) + 0.0105 * glucose) : null
        }
      case RaceEthnicity.White:
        // Adult Male White Non-Hispanic
        return {
          mets_z_bmi: bmi ? -4.8316 + 0.0315 * bmi - 0.0272 * hdl + 0.0044 * sbp + 0.8018 * log(triglyceride) + 0.0101 * glucose : null,
          mets_z_wc: waist ? (-5.4559 + 0.0125 * waist - 0.0251 * hdl + 0.0047 * sbp + 0.8244 * log(triglyceride) + 0.0106 * glucose) : null
        }
    }
  }

  /* Adolescents */
  if (sex === Sex.Female) {
    switch (race) {
      case RaceEthnicity.Black:
        return {
          mets_z_bmi: bmiZScore ? -3.7145 + 0.5136 * bmiZScore - 0.0190 * hdl + 0.0131 * sbp + 0.4442 * log(triglyceride) + 0.0108 * glucose : null
        }
      case RaceEthnicity.Hispanic:
        return {
          mets_z_bmi: bmiZScore ? -4.7637 + 0.3520 * bmiZScore - 0.0263 * hdl + 0.0152 * sbp + 0.6910 * log(triglyceride) + 0.0133 * glucose : null
        }
      case RaceEthnicity.White:
        return {
          mets_z_bmi: bmiZScore ? -4.3757 + 0.4849 * bmiZScore - 0.0176 * hdl + 0.0257 * sbp + 0.3172 * log(triglyceride) + 0.0083 * glucose : null
        }
    }
  }

  /* Male */
  switch (race) {
    case RaceEthnicity.Black:
      return {
        mets_z_bmi: -4.7544 + 0.2401 * bmiZScore - 0.0284 * hdl + 0.0134 * sbp + 0.6773 * log(triglyceride) + 0.0179 * glucose
      }
    case RaceEthnicity.Hispanic:
      return {
        mets_z_bmi: -3.2971 + 0.2930 * bmiZScore - 0.0315 * hdl + 0.0109 * sbp + 0.6137 * log(triglyceride) + 0.0095 * glucose
      }
    case RaceEthnicity.White:
      return {
        mets_z_bmi: -4.931 + 0.2804 * bmiZScore - 0.0257 * hdl + 0.0189 * sbp + 0.6240 * log(triglyceride) + 0.0140 * glucose
      }
  }
}

// Percentile calculates the percentile from the specified z-score, z. This can
// be used for both the BMI and MetS percenitles.
// https://github.com/travm/ZMI/blob/5340e9d194cc716dd4cf9ff7b5ae479e915798a5/app/js/zmi.js +194
function Percentile (z) {
  return 100 * (1 / (1 + Math.exp(-0.07056 * Math.pow(z, 3) - (1.5976 * z))))
}

module.exports = {
  CalculateMSS,
  Percentile,
  RaceEthnicity,
  Sex
}


src/web.js
/*
   Copyright 2019 University of Florida

   Licensed under the Apache License, Version 2.0 (the "License");
   you may not use this file except in compliance with the License.
   You may obtain a copy of the License at

       http://www.apache.org/licenses/LICENSE-2.0

   Unless required by applicable law or agreed to in writing, software
   distributed under the License is distributed on an "AS IS" BASIS,
   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
   See the License for the specific language governing permissions and
   limitations under the License.
*/

var React = window.React || require('react')
var ReactDOM = window.ReactDOM || require('react-dom')
var moment = window.moment || require('moment')

const msscalc = require('./msscalc')
const bmi = require('./bmi')

class Calculator extends React.Component {
  constructor (props) {
    super(props)

    this.state = {
      age: null,
      sex: '',
      race: '',
      hdl: '',
      sbp: '',
      triglyceride: '',
      glucose: '',
      waist: '',
      waistUnit: 'cm',
      weight: '',
      weightUnit: 'kg',
      height: '',
      heightUnit: 'cm',
      bmiadult: '',
      bmiz: '',
      birth: '',
      appointment: moment().format(moment.HTML5_FMT.DATE),
      result: null
    }

    this.handleBack = this.handleBack.bind(this)
    this.handleChange = this.handleChange.bind(this)
    this.handleClick = this.handleClick.bind(this)
    this.handleSubmit = this.handleSubmit.bind(this)
    this.afterUpdate = this.afterUpdate.bind(this)
  }

  render () {
    let {
      age, sex, race, weight, weightUnit, height, heightUnit, hdl, sbp, triglyceride,
      glucose, waist, waistUnit, bmiadult, bmiz, birth, appointment, result
    } = this.state

    const adolescent = age !== null && age < 20

    return (
      <form onSubmit={this.handleSubmit}>
        <h3>Demographics</h3>
        <div className="form-group">
          <label htmlFor="birth">Birthdate <em>(if younger than 20 years old)</em></label>
          <input className="form-control" type="date" name="birth" value={birth}
            placeholder="Ex. 1984-12-23"
            onChange={this.handleChange}></input>
        </div>
        {birth && (
          <div className="form-group">
            <label htmlFor="appointment">Appointment Date</label>
            <input className="form-control" type="date" name="appointment" value={appointment} onChange={this.handleChange}></input>
          </div>
        )}
        {birth && appointment && (
          <div className="form-group">
            <label htmlFor="age">Age at Appointment (years)</label>
            <input className="form-control" name="age" type="number" readOnly
              value={age === null ? '' : age}
            />
          </div>
        )}

        <ButtonGroup
          name="sex" label="Sex" value={sex} required options={msscalc.Sex}
          onClick={this.handleClick}
        />

        <ButtonGroup
          name="race" label="Race and Ethnicity" value={race} required
          options={{
            'Hispanic': msscalc.RaceEthnicity.Hispanic,
            'Non-Hispanic Black': msscalc.RaceEthnicity.Black,
            'Non-Hispanic White': msscalc.RaceEthnicity.White
          }}
          onClick={this.handleClick}
        />

        <h3>Measurements</h3>
        <div className="form-group">
          <label htmlFor="height">Height</label>
          <Measurement name="height" value={height} unit={heightUnit}
            min="0" required
            onValueChange={this.handleChange} onUnitChange={this.handleChange}
            units={{
              cm: 'Centimeters (cm)',
              in: 'Inches (in)'
            }}
          />
        </div>

        <label htmlFor="weight">Weight</label>
        <Measurement name="weight" value={weight} unit={weightUnit}
          min="0" required
          onValueChange={this.handleChange} onUnitChange={this.handleChange}
          units={{
            kg: 'Kilograms (kg)',
            lbs: 'Pounds (lbs)'
          }}
        />

        {!adolescent && (
          <div className="form-group">
            <label htmlFor="waist">Waist Circumference <em>(if available)</em></label>
            <Measurement name="waist" value={waist} unit={waistUnit}
              min="0"
              onValueChange={this.handleChange} onUnitChange={this.handleChange}
              units={{
                cm: 'Centimeters (cm)',
                in: 'Inches (in)'
              }}
            />
          </div>
        )}

        {bmiadult && (
          <div className="form-group">
            <label htmlFor="bmiadult"> BMI </label>
            <input className="form-control" name="bmiadult" value={bmiadult.toFixed(3)} readOnly />
          </div>
        )}

        <div className="form-group">
          <label htmlFor="sbp">Systolic Blood Pressure (mmHg)</label>
          <input className="form-control" name="sbp" type="number"
            min="0" required step="any" value={sbp}
            placeholder="Ex: 120"
            onChange={this.handleChange}></input>
        </div>

        <h3>Lab Values</h3>
        <div className="form-group">
          <label htmlFor="hdl"><abbr title="High-density lipoprotein">HDL</abbr> (mg/dL)</label>
          <input className="form-control" name="hdl" type="number" required
            min="0" step="any" value={hdl}
            placeholder="Ex: 50"
            onChange={this.handleChange}></input>
        </div>

        <div className="form-group">
          <label htmlFor="triglyceride">Triglycerides (mg/dL)</label>
          <input className="form-control" name="triglyceride" type="number"
            min="0" step="any" required value={triglyceride}
            placeholder="Ex: 120"
            onChange={this.handleChange}></input>
        </div>

        <div className="form-group">
          <label htmlFor="glucose">Fasting Glucose (mg/dL)</label>
          <input className="form-control" name="glucose" type="number"
            min="0" step="any" required value={glucose}
            placeholder="Ex: 75"
            onChange={this.handleChange}></input>
        </div>

        {!result && (
          <button type="submit" className="btn btn-primary float-right">
          Calculate
          </button>
        )}

        {result && (
          <div className="result">
            <h2>Results</h2>

            <div className="row">
              {adolescent && bmiz && (
                <div className="col-sm">
                  <p>BMI Z-Score for Adolescents
                    <span className="amount">{bmiz.toFixed(3)}</span>
                  </p>
                  <p>BMI Percentile for Adolescents
                    <span className="amount">{msscalc.Percentile(bmiz).toFixed(2)}%</span>
                  </p>
                </div>
              )}

              <div className="col-sm">
                {result.mets_z_bmi && (
                  <p>MetS Z-Score based on Body Mass Index
                    <span className="amount">{result.mets_z_bmi.toFixed(3)}</span>
                  </p>
                )}
                {result.mets_z_bmi && (
                  <p>MetS Percentile based on Body Mass Index
                    <span className="amount">{msscalc.Percentile(result.mets_z_bmi).toFixed(2)}%</span>
                  </p>
                )}
              </div>

              {result.mets_z_wc && (
                <div className="col-sm">
                  {result.mets_z_wc && (
                    <p>MetS Z-Score based on Waistline
                      <span className="amount">{result.mets_z_wc.toFixed(3)}</span>
                    </p>
                  )}
                  {result.mets_z_wc && (
                    <p>MetS Percentile based on Waistline
                      <span className="amount">{msscalc.Percentile(result.mets_z_wc).toFixed(2)}%</span>
                    </p>
                  )}
                </div>
              )}

            </div>
          </div>
        )}

      </form>
    )
  }

  handleBack (event) {
    event.preventDefault()
    this.setState({ result: null })
  }

  handleChange (event) {
    event.persist()

    this.setState({ [event.target.name]: event.target.value }, () => {
      this.afterUpdate()
      runValidator(event.target)
    })
  }

  handleClick (event) {
    let input

    switch (event.target.tagName.toLowerCase()) {
      case 'input':
        input = event.target
        break

      case 'label':
        input = event.target.getElementsByTagName('input')[0]
        break
    }

    if (!input) {
      return
    }

    this.setState({ [input.name]: input.value }, () => {
      this.afterUpdate()
      runValidator(input)
    })
  }

  handleSubmit (event) {
    event.preventDefault()

    let {
      age, bmiadult, sex, race, hdl, sbp, triglyceride, glucose, waist, waistUnit,
      birth, appointment, bmiz, weight, height
    } = this.state

    if (!(sex && race && sbp && glucose && hdl && triglyceride) || !((weight && height) || waist)) {
      alert('Please fill out the required fields.')
      return
    }

    if (age !== null && age < 2) {
      alert('The age at appointment time must be at least 2 years old.')
      return
    }

    const result = msscalc.CalculateMSS({
      age: age ? moment(appointment).diff(moment(birth), 'years') : 25,
      sex,
      race,
      bmi: bmiadult,
      hdl,
      sbp,
      triglyceride,
      glucose,
      bmiZScore: bmiz,
      waist: waist && (age === null || age > 20) ? centimeters(waist, waistUnit) || null : null
    })

    this.setState({ result })
  }

  afterUpdate () {
    const { birth, appointment, weight, weightUnit, height, heightUnit, sex } = this.state

    let age = null
    let bmiz = ''
    let bmiadult = ''

    if (birth && appointment) {
      age = moment(appointment).diff(moment(birth), 'years')
    }

    if (height && weight) {
      const weightKG = kilograms(weight, weightUnit)
      const heightMeters = meters(height, heightUnit)

      bmiadult = bmi.BMIAdult(weightKG, heightMeters)

      const adolescent = age !== null && age >= 2 && age < 20
      if (adolescent && sex) {
        const agemos = moment(appointment).diff(moment(birth), 'months')
        const sexord = sex === 'MALE' ? bmi.Sex.Male : bmi.Sex.Female
        bmiz = bmi.BMIZscore(weightKG, heightMeters, sexord, agemos)
      }
    }

    this.setState({
      age,
      bmiz,
      bmiadult: bmiadult,
      result: null
    })
  }
}

// meters converts the length from units into meters.
function meters (length, units) {
  if (isNaN(parseFloat(length))) {
    console.error('length must be a number; got:', length)
    return null
  }

  switch (units) {
    case 'm':
      return length

    case 'cm':
      return length / 100

    case 'in':
      // https://www.google.com/search?q=in+to+m
      return length / 39.37
  }

  console.error("units must be 'm', 'cm', or 'in'; got:", units)
  return null
}

function centimeters (length, units) {
  if (isNaN(parseFloat(length))) {
    console.error('length must be a number; got:', length)
    return null
  }

  switch (units) {
    case 'cm':
      return length

    case 'm':
      return length * 100

    case 'in':
      // https://www.google.com/search?q=in+to+cm
      return length * 2.54
  }

  console.error("units must be 'm', 'cm', or 'in'; got:", units)
  return null
}

// kilograms converts the mass from units into kilograms.
function kilograms (mass, units) {
  if (isNaN(parseFloat(mass))) {
    console.error('mass must be a number; got:', mass)
    return null
  }

  switch (units) {
    case 'kg':
      return mass

    case 'lbs':
      // See https://www.ngs.noaa.gov/PUBS_LIB/FedRegister/FRdoc59-5442.pdf
      return mass * 0.45359237
  }

  console.error("units must be 'kg' or 'lbs'; got:", units)
  return null
}

function runValidator (target) {
  if (!target.checkValidity || !target.reportValidity) {
    return
  }

  if (!target.checkValidity()) {
    target.classList.add('is-invalid')
  } else {
    target.classList.remove('is-invalid')
  }
  target.reportValidity()
}

function ButtonGroup (props) {
  const { name, label, options, value, required, onClick, onKeyPress } = props

  if (!options || options.length === 0) {
    return null
  }

  return (
    <div className="form-group">
      <label htmlFor={name}>{label}</label>
      <div className="input-group btn-group btn-group-toggle" data-toggle="buttons">
        {Object.keys(options).map(label => (
          <Button key={label}
            required={required}
            group={name} label={label} value={options[label]}
            pressed={value === options[label]}
            onClick={onClick} onKeyPress={onKeyPress}
          />
        ))}
      </div>
    </div>
  )
}

function Button (props) {
  const { group, label, pressed, value, onClick, onKeyPress, required = false } = props

  return (
    <label className= {`btn btn-light ${pressed ? 'active' : ''}`}
      role="button"
      checked={pressed}
      aria-pressed={pressed}
      onClick={onClick} onKeyPress={onKeyPress}
    >
      <input type="radio" name={group} value={value} autoComplete="off" required={required} /> {label}
    </label>
  )
}

function Measurement (props) {
  const {
    name, unit, units, value, onValueChange, onUnitChange, min, max, required
  } = props

  return (
    <div className="Measurement input-group">
      <input className="form-control" name={name} type="number" min={min} max={max}
        required={required} step="any" value={value} onChange={onValueChange}></input>
      {units && (
        <div className="input-group-append">
          <select className="custom-select" value={unit} name={`${name}Unit`} onChange={onUnitChange}>
            {Object.keys(units).map(key => (
              <option key={key} value={key}>{units[key]}</option>
            ))}
          </select>
        </div>
      )}
    </div>
  )
}

ReactDOM.render(<Calculator />, document.getElementById('calculator'))

function resizeBtnGroup () {
  const groups = document.getElementsByClassName('btn-group')

  const smallWindow = window.innerWidth < 500
  for (let i = 0; i < groups.length; i++) {
    const group = groups[i]
    if (smallWindow) {
      group.classList.add('btn-group-vertical')
    } else {
      group.classList.remove('btn-group-vertical')
    }
  }
}

window.onresize = resizeBtnGroup
resizeBtnGroup()
